src/overlays/actors/ovl_Elf_Msg2/z_elf_msg2.c

```diff
@@ -21,23 +50,177 @@ const ActorInit Elf_Msg2_InitVars = {
+        LOG_STRING("共倒れ", "../z_elf_msg2.c", 192);
+        Actor_Kill(&this->actor);
+        return 1;
+    }
+    return 0;
+}
+
+void ElfMsg2_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ElfMsg2* this = THIS;
+
+    osSyncPrintf(VT_FGCOL(CYAN) " Elf_Msg2_Actor_ct %04x\n\n" VT_RST, this->actor.params);
+    if (!ElfMsg2_KillCheck(this, globalCtx)) {
+        if ((this->actor.posRot.rot.x > 0) && (this->actor.posRot.rot.x < 8)) {
+            this->actor.unk_1F = this->actor.posRot.rot.x - 1;
+        }
+        Actor_ProcessInitChain(thisx, sInitChain);
+        if (this->actor.posRot.rot.y >= 0x41) {
+            ElfMsg2_SetupAction(this, ElfMsg2_Idle);
+        } else {
+            ElfMsg2_SetupAction(this, ElfMsg2_WaitForRead);
+            this->actor.flags |= 0x00040001; // Make actor targetable and Navi-checkable
+            this->actor.textId = ElfMsg2_GetMessageId(this);
+        }
+        this->actor.shape.rot.x = this->actor.shape.rot.y = this->actor.shape.rot.z = 0;
+    }
+}
+
+void ElfMsg2_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/ElfMsg2_Init.s")
+s32 ElfMsg2_GetMessageId(ElfMsg2* this) {
+    return (this->actor.params & 0xFF) + 0x100;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/ElfMsg2_Destroy.s")
+/**
+ * Runs while Navi text is up. Kills the actor upon closing the text box unless rot.z == 1, can also set a switch flag
+ * from params.
+ */
+void ElfMsg2_Read(ElfMsg2* this, GlobalContext* globalCtx) {
+    s32 switchFlag;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/func_809AD968.s")
+    // If the text box is closing this frame:
+    if (func_8002F334(&this->actor, globalCtx)) {
+        if (this->actor.posRot.rot.z != 1) {
+            Actor_Kill(&this->actor);
+            switchFlag = (this->actor.params >> 8) & 0x3F;
+            if (switchFlag != 0x3F) {
+                Flags_SetSwitch(globalCtx, switchFlag);
+            }
+        } else {
+            ElfMsg2_SetupAction(this, ElfMsg2_WaitForRead);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/func_809AD978.s")
+/**
+ * Runs while Navi text is not up. If text box flag is set, this function unsets it and proceeds to ElfMsg2_Read.
+ */
+void ElfMsg2_WaitForRead(ElfMsg2* this, GlobalContext* globalCtx) {
+    if (func_8002F194(&this->actor, globalCtx)) {
+        ElfMsg2_SetupAction(this, ElfMsg2_Read);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/func_809AD9F4.s")
+/**
+ * Idles until a switch flag is set, at which point the actor becomes interactable
+ */
+void ElfMsg2_Idle(ElfMsg2* this, GlobalContext* globalCtx) {
+    // If (y < 0x41) && (y < 0x81), Idles until switch flag (actor.posRot.rot.y - 0x41) is set
+    // If (y > 0x80), Idles forever (Bug?)
+    if ((this->actor.posRot.rot.y >= 0x41) && (this->actor.posRot.rot.y < 0x81) &&
```

Comment:
Very minor but I'd say these numbers make more sense (also relevant in other places)
```suggestion
    // If (y < 0x41) && (y < 0x80), Idles until switch flag (actor.posRot.rot.y - 0x41) is set
    // If (y > 0x80), Idles forever (Bug?)
    if ((this->actor.posRot.rot.y >= 0x41) && (this->actor.posRot.rot.y <= 0x80) &&
```